src/GithubApi.hs

```diff
@@ -180,23 +180,19 @@ runGithub auth projectInfo operation =
 -- the sense of being observable by Github users. We will still make requests
 -- against the read-only endpoints of the API. This is useful for local testing.
 runGithubReadOnly
-  :: MonadIO m
-  => MonadLogger m
+  :: (IOE :> es, MonadLoggerEffect :> es)
   => Github3.Auth
   -> ProjectInfo
-  -> GithubOperationFree a
-  -> m a
-runGithubReadOnly auth projectInfo operation =
-  let
-    unsafeResult = runGithub auth projectInfo operation
-  in
-    case operation of
+  -> Eff (GithubOperation : es) a
+  -> Eff es a
+runGithubReadOnly auth projectInfo = runGithub auth projectInfo . augmentedGithubOperation
+  where
+    augmentedGithubOperation = interpose $ \_ operation -> case operation of
       -- These operations are read-only, we can run them for real.
-      HasPushAccess {} -> unsafeResult
-      GetPullRequest {} -> unsafeResult
-      GetOpenPullRequests {} -> unsafeResult
+      HasPushAccess username -> send $ HasPushAccess username
+      GetPullRequest pullRequestId -> send $ GetPullRequest pullRequestId
+      GetOpenPullRequests -> send GetOpenPullRequests
```

Comment:
Why did you choose to change the style here and go for explicit calls?

Reply:
I still think that with a let-binding, it could be more practical than the current choice. It would make clearer that we're just sending the exact same stuff to `runGithub`.

The choice you made looks nicer to the eye, but doesn't express exactly the same thing, even if the overall behaviour is the same.

Reply:
We discussed this offline. It turns out that `localSeqUnlift` is subject to some rather quirky type behavior. For example, moving the `localSeqUnlift ...` expression to a `let`-binding right before the `case` leads to the compiler no longer agreeing that we're making a valid move. This is because the `case` is on a constructors of a GADT, which gives the compiler extra typing information. Since the let-binding would be outside of the `case`, that information is no longer there, and it would have to be passed explicitly.

This is done in the solution from @DiegoDiverio:

```haskell
runGithubReadOnly
  :: (IOE :> es, MonadLoggerEffect :> es)
  => Github3.Auth
  -> ProjectInfo
  -> Eff (GithubOperation : es) a
  -> Eff es a
runGithubReadOnly auth projectInfo = runGithub auth projectInfo . augmentedGithubOperation
  where
    liftLocal localEnv op = localSeqUnlift localEnv $ \unlift -> unlift $ send op
    augmentedGithubOperation = interpose $ \localEnv operation -> case operation of
      -- These operations are read-only, we can run them for real.
      HasPushAccess {} -> liftLocal localEnv operation
      GetPullRequest {} -> liftLocal localEnv operation
      GetOpenPullRequests {} -> liftLocal localEnv operation

      -- These operations have side effects, we fake them.
      LeaveComment pr body ->
        logInfoN $ format "Would have posted comment on {}: {}" (show pr, body)
```

There is an argument to be made that having the same text on each line is an improvement. Though in my opinion, the added complexity is no longer worth the reading benefit. I'll leave it as it is :+1: